bwe/BandwidthEstimator.cpp

```diff
@@ -31,164 +43,243 @@ BandwidthEstimator::BandwidthEstimator(const Config& config)
     _covarianceP = initDelta * math::transpose(initDelta);
 }
 
 void BandwidthEstimator::onUnmarkedTraffic(uint32_t packetSize, uint64_t receiveTimeNs)
 {
-    if (_baseClockOffset != 0 && _state(0) < 1600 * 8)
+    if (_baseClockOffset != 0 && _state(QueuedBits) < _config.mtu * 2 * 8)
     {
-        _state(0) = _state(0) + packetSize * 8;
+        _state(QueuedBits) += packetSize * 8;
     }
     _previousReceiveTime = receiveTimeNs;
-    _receiveBandwidth.update(packetSize * 8, receiveTimeNs);
+    _receiveBitrate.update(packetSize * 8, receiveTimeNs);
+}
+
+double BandwidthEstimator::predictDelay() const
+{
+    const double offsetAdjustment = _packetSize0 * 8 / _state(Bandwidth);
+    return predictAbsoluteDelay(_state) - _state(ClockOffset) + offsetAdjustment;
 }
 
 void BandwidthEstimator::update(uint32_t packetSize, uint64_t transmitTimeNs, uint64_t receiveTimeNs)
 {
-    if (_baseClockOffset == 0 && _state(2) == 0 && _state(0) == 0 && _previousTransmitTime == 0)
+    assert(_state(Bandwidth) > 10);
+    if (_baseClockOffset == 0 && _state(QueuedBits) == 0 && _previousTransmitTime == 0)
     {
         // base Offset is very sensitive, if you start 5ms behind it will create a lower estimate, assuming higher delay
         // and longer queue. Starting with a long queue is also a bad start
         _baseClockOffset = receiveTimeNs - transmitTimeNs;
         _previousTransmitTime = transmitTimeNs - 5 * utils::Time::sec;
         _previousReceiveTime = receiveTimeNs - 5 * utils::Time::ms;
+        _packetSize0 = packetSize;
     }
 
     const double tau = std::max(0.0, static_cast<double>(transmitTimeNs - _previousTransmitTime) / utils::Time::ms);
     const double observedDelay =
         static_cast<double>(static_cast<int64_t>(receiveTimeNs - transmitTimeNs - _baseClockOffset)) / utils::Time::ms;
 
-    const auto actualDelay = (observedDelay - _state(2));
-    const double congestionScale = analyseCongestion(actualDelay, packetSize, receiveTimeNs);
+    const auto expectedState = transitionState(packetSize, tau, _state);
+    const auto expectedDelay = predictAbsoluteDelay(expectedState);
+    _congestion.countDelays(observedDelay - expectedDelay);
 
-    _receiveBandwidth.update(packetSize * 8, receiveTimeNs);
-    // predict mean state
+    auto actualDelay = (observedDelay - _state(ClockOffset));
+    if (actualDelay < 0)
+    {
+        _state(QueuedBits) = 0; // queue must be empty before this packet
+        _state(ClockOffset) = observedDelay;
+        actualDelay = 0;
+        _packetSize0 = packetSize;
+    }
+    else if (actualDelay == 0 && packetSize > _packetSize0)
+    {
+        _packetSize0 = packetSize;
+    }
+    else if (observedDelay > expectedDelay && tau > 20.0 && expectedState(QueuedBits) < _config.mtu * 8 * 3)
+    {
+        // inter transmit interval larger than 20ms risk having radio shut down. adjust co up
```

Comment:
For LTE network, the radio can be put to sleep when there is a transmission pause. It does not care that it is running behind. This allows bwe to measure observed throughput. Improvement is +1Mbps.
It does not allow major increase, which could lead to over estimation after pause on back bone link. It merly makes the first packet fit current estimate. If this is wrong assumption the clock offset will quickly be adjusted down. Note that we require queue to be drained, and that is when this matters most.